draco/asp/hard.lp

```diff
@@ -2,10 +2,48 @@
 
 % @hard(text_mark_without_text_channel) Text mark requires text encoding.
 violation(text_mark_without_text_channel) :-
-    attribute(mark_type,M,text),
+    attribute((mark,type),M,text),
     not attribute(has_channel,M,text).
 
 % @hard(text_channel_without_text_mark) Text channel requires text mark.
 violation(text_channel_without_text_mark) :-
-    not attribute(mark_type,M,text),
+    not attribute((mark,type),M,text),
     attribute(has_channel,M,text).
+
+% @hard(bin_and_aggregate) Cannot bin and aggregate.
+violation(bin_and_aggregate) :-
+    attribute((encoding,binning),E,_),
+    attribute((encoding,aggregate),E,_).
+
+% @hard(no_encodings) There has to be at least one encoding.
+violation(no_encodings) :- not entity(encoding,_,_).
+
+% @hard(repeat_channel) Cannot use single channels twice.
+violation(repeat_channel) :-
```

Comment:
I think this is wrong now since we can actually reuse the same encoding in differnet marks or different views.

Reply:
modified